apps/web/src/app/(profile)/[username]/_components/dashboard/_actions.ts

```diff
@@ -167,72 +173,34 @@ export async function getSolvedChallenges(userId: string) {
   };
 }
 
-export interface BadgeInfo {
-  // eslint-disable-next-line @typescript-eslint/sort-type-constituents
-  slug: 'aot-2023-bronze' | 'aot-2023-silver' | 'aot-2023-gold' | 'aot-2023-platinum';
+export type AotBadges =
+// eslint-disable-next-line @typescript-eslint/sort-type-constituents
+  | 'aot-2023-bronze'
+  | 'aot-2023-silver'
+  | 'aot-2023-gold'
+  | 'aot-2023-platinum';
+// eslint-disable-next-line @typescript-eslint/sort-type-constituents
+export type BadgeLevels = 'bronze' | 'silver' | 'gold' | 'platinum';
+
+export interface Badges<T> {
+  slug: T;
   name: string;
+  shortName: string;
 }
 
-export async function getBadges(userId: string): Promise<BadgeInfo[]> {
-  const badges: BadgeInfo[] = [];
+export type AllBadges = Badges<AotBadges | BadgeLevels>;
 
-  const holidayTrack = await prisma.track.findFirst({
-    where: {
-      slug: 'advent-of-typescript-2023',
-    },
-    include: {
-      trackChallenges: {
-        orderBy: {
-          orderId: 'asc',
-        },
-        include: {
-          challenge: {
-            include: {
-              submission: {
-                where: {
-                  userId,
-                },
-              },
-            },
-          },
-        },
-      },
-      enrolledUsers: {
-        where: {
-          id: userId,
-        },
-        select: {
-          id: true,
-        },
-      },
-    },
-  });
-
-  const numberOfAttemptedHolidayChallenges =
-    holidayTrack?.trackChallenges.filter((trackChallenge) => {
-      return (trackChallenge.challenge.submission?.length ?? 0) > 0;
-    }).length ?? 0;
-
-  if (numberOfAttemptedHolidayChallenges > 0) {
-    badges.push({ slug: 'aot-2023-bronze', name: 'Advent of TypeScript 2023 Bronze' });
-  }
-
-  const numberOfCompletedHolidayChallenges =
-    holidayTrack?.trackChallenges.filter((trackChallenge) => {
-      return trackChallenge.challenge.submission?.some((submission) => submission.isSuccessful);
-    }).length ?? 0;
-
-  if (numberOfCompletedHolidayChallenges >= 5) {
-    badges.push({ slug: 'aot-2023-silver', name: 'Advent of TypeScript 2023 Silver' });
-  }
+export async function getBadges(userId: string): Promise<AllBadges[]> {
+  const badges: AllBadges[] = [];
```

Comment:
im not seeing where we handle "events"

Reply:
I have to make another table that awards the badges when the criteria is met. I think it would be awarded in the feed. Hmmm good comment, give me a bit.

Reply:
why do we need another table. i thought we agreed we just compute it

Reply:
Do these code comments make sense? basically compute the current badges, then afterwards award in the feed for display what badges you have.

Is the latter part not needed yet until the feed is built out? That seems like a relatively large feature?

```ts

  // Retrieve current data from db
  const advent: AdventChallenges | null = await AdventRetrieveData(userId);
  const difficulty: Difficulty[] | null = await DifficultyRetrieveData(userId);
  const sharedSolutions: SharedTotals[] | null = await SharedSolutionRetrieveData(userId);

  // determine badges user should have
  await AdventChallengeFn(currentBadgeStatus, advent);
  await DifficultyBadgesFn(currentBadgeStatus, difficulty ?? []);
  await SharedBadgesFn(currentBadgeStatus, sharedSolutions ?? []);

  // TODO: Retrieve Awarded Badges From Feed

  // TODO: Award Missing Badges into Feed Table

```

Reply:
The above was filled out to recalculate badges on each event.

I've commented elsewhere where the events are triggered to recalculate and award badges. Comment there for now if you disagree with the placement or logic.